ddtrace/llmobs/_evaluators/ragas/faithfulness.py

```diff
@@ -13,24 +65,236 @@ class RagasFaithfulnessEvaluator:
+        context - input.prompt.variables.context
+        answer - output.messages[-1].content
+        """
+        with self.llmobs_service.workflow("ragas.extract_faithfulness_inputs"):
+            self.llmobs_service.annotate(input_data=span_event)
+            question, answer, context = None, None, None
+
+            meta_io = span_event.get("meta")
+            if meta_io is None:
+                return None
+
+            meta_input = meta_io.get("input")
+            meta_output = meta_io.get("output")
+
+            if not (meta_input and meta_output):
+                return None
+
+            prompt = meta_input.get("prompt")
+            if prompt is None:
+                return None
+            prompt_variables = prompt.get("variables")
+
+            input_messages = meta_input.get("messages")
+
+            messages = meta_output.get("messages")
+            if messages is not None and len(messages) > 0:
+                answer = messages[-1].get("content")
+
+            if prompt_variables:
+                question = prompt_variables.get("question")
+                context = prompt_variables.get("context")
+
+            if not question and len(input_messages) > 0:
+                question = input_messages[-1].get("content")
+
+            self.llmobs_service.annotate(output_data={"question": question, "context": context, "answer": answer})
+            if any(field is None for field in (question, context, answer)):
+                logger.debug("Failed to extract inputs required for faithfulness evaluation")
+                return None
+
+            return {
+                "question": question,
+                "context": context,
+                "answer": answer,
+            }
+
+    def _create_statements_prompt(self, answer, question):
+        with self.llmobs_service.task("ragas.create_statements_prompt"):
+            sentences = self.split_answer_into_sentences.segment(answer)
+            sentences = [sentence for sentence in sentences if sentence.strip().endswith(".")]
+            sentences = "\n".join([f"{i}:{x}" for i, x in enumerate(sentences)])
+            return self.ragas_faithfulness_instance.statement_prompt.format(
+                question=question, answer=answer, sentences=sentences
+            )
+
+    def _create_natural_language_inference_prompt(self, statements, context_str):
+        with self.llmobs_service.task("ragas.create_natural_language_inference_prompt"):
+            statements_str: str = json.dumps(statements)
+            prompt_value = self.ragas_faithfulness_instance.nli_statements_message.format(
+                context=context_str, statements=statements_str
+            )
+            return prompt_value
+
+    def _compute_score(self, answers):
+        with self.llmobs_service.task("ragas.compute_score"):
+            faithful_statements = sum(1 if answer.verdict else 0 for answer in answers.__root__)
```

Comment:
I see we're using quite a few instances of private attributes/fields from RAGAS objects. Is this publicly documented to use them this way? If not, should we be careful about using them in case they get changed?

Reply:
Good point; i think all the attributes accessed are public except for `faithfulness._reproducibility` (this determines `n` parameter for the llm call. 

I added a check to make sure that attribute is present, else default to 1 if not.

`reproducibility = self.ragas_faithfulness_instance._reproducibility if hasattr(self.ragas_faithfulness_instance, "_reproducibility") else 1
`

I think one difficult part is parsing the inputs and outputs to llm's calls ragas makes. `models.py` contains I/O models defined by us (similar code also exists in the ragas python lib, but is private) which specify the expected input format & output format to LLMs. So were technically not accessing private fields, but there's some compatibility issues here:
- If the user uses a custom prompt that deviates from this I/O structure, it'll break the faithfulness generation, but AFAIK this is an inherent limitation of the ragas library
- If ragas changes their default prompt for llm generations to break from the models that we've defined, parsing will fail for the integration.
    -  I think the better approach here is to try to access their private I/O models, and if we can't fallback to default ones we've defined. _**Is it ok if I do this in a follow up PR?**_

Reply:
Fine with me to do this in a follow-up PR